tools/omnissm/functions/handle-registrations/main.go

```diff
@@ -34,76 +32,41 @@ type registrationHandler struct {
 	logger.Info().Interface("identity", req.Identity()).Msg("new registration request")
-	entry, err, ok := r.OmniSSM.Registrations.Get(ctx, req.Identity().Hash())
+	resp, err := r.OmniSSM.RequestActivation(ctx, req)
 	if err != nil {
 		return nil, err
 	}
-	if ok {
-		logger.Info().Interface("entry", entry).Msg("existing registration entry found")
-		return &omnissm.RegistrationResponse{RegistrationEntry: *entry, Region: req.Identity().Region}, nil
+	if resp.Existing() {
+		logger.Info().Interface("entry", resp).Msg("existing registration entry found")
+	} else {
+		logger.Info().Interface("entry", resp).Msg("new registration entry created")
 	}
-	activation, err := r.SSM.CreateActivation(ctx, req.Identity().Name())
-	if err != nil {
-		if r.OmniSSM.SQS != nil && request.IsErrorThrottle(err) || request.IsErrorRetryable(err) {
-			sqsErr := r.OmniSSM.SQS.Send(ctx, &omnissm.DeferredActionMessage{
-				Type:  omnissm.CreateActivation,
-				Value: req.Identity(),
-			})
-			if sqsErr != nil {
-				return nil, sqsErr
-			}
-			return nil, errors.Wrapf(err, "deferred action to SQS queue: %#v", r.OmniSSM.Config.QueueName)
-		}
-		return nil, err
-	}
-	entry = &omnissm.RegistrationEntry{
-		Id:         req.Identity().Hash(),
-		CreatedAt:  time.Now().UTC(),
-		AccountId:  req.Identity().AccountId,
-		Region:     req.Identity().Region,
-		InstanceId: req.Identity().InstanceId,
-		Activation: *activation,
-		ManagedId:  "-",
-	}
-	if err := r.OmniSSM.Registrations.Put(ctx, entry); err != nil {
-		if r.OmniSSM.SQS != nil && request.IsErrorThrottle(err) || request.IsErrorRetryable(err) {
-			sqsErr := r.OmniSSM.SQS.Send(ctx, &omnissm.DeferredActionMessage{
-				Type:  omnissm.PutRegistrationEntry,
-				Value: entry,
-			})
-			if sqsErr != nil {
-				return nil, sqsErr
-			}
-			return nil, errors.Wrapf(err, "deferred action to SQS queue: %#v", r.OmniSSM.Config.QueueName)
-		}
-		return nil, err
-	}
-	logger.Info().Interface("entry", entry).Msg("new registration entry created")
-	return &omnissm.RegistrationResponse{RegistrationEntry: *entry, Region: req.Identity().Region}, nil
+	return resp, nil
 }
 
 func (r *registrationHandler) UpdateRegistration(ctx context.Context, req *omnissm.RegistrationRequest) (*omnissm.RegistrationResponse, error) {
 	logger := log.With().Str("handler", "UpdateRegistration").Logger()
 	logger.Info().Interface("identity", req.Identity()).Msg("update registration request")
+	if !ssm.IsManagedInstance(req.ManagedId) {
+		return nil, lambda.BadRequestError{fmt.Sprintf("invalid managedId %#v", req.ManagedId)}
+	}
 	id := req.Identity().Hash()
 	entry, err, ok := r.OmniSSM.Registrations.Get(ctx, id)
 	if err != nil {
 		return nil, err
 	}
 	if !ok {
 		logger.Info().Str("instanceName", req.Identity().Name()).Str("id", id).Msg("registration entry not found")
-		return nil, errors.Wrapf(err, "entry not found: %#v", id)
+		return nil, lambda.NotFoundError{"entry not found: " + id}
```

Comment:
for these could we retain `%#v` so that the string zero value is better represented in returned errors?

Reply:
agreed, fixed.